### PR TITLE
Update key generation to provide fixed signatures and encoded checksums (for improved security tooling)

### LINKED ISSUE #7662: Consider updating key generation to increase identifiability (and glean security value from improved tooling)
#### What problem would the feature you're requesting solve? Please describe.
Inadvertent plaintext exposure of secrets is a significant industry problem. Many modern security models are therefore updating their key generation logic to encode fixed signatures and checksums in secret data. This approach allows various scanners (such as Microsoft Defender and [GitHub Token Scanning](https://github.blog/2018-10-17-behind-the-scenes-of-github-token-scanning/), to efficiently detect/block exposures.

Without improvement of this kind, tools must either generate large #'s of false positives (which prevent meaningful engineer review and security response) or favor approaches that generate false negatives (which leaves projects exposed to security risk).

#### Describe the solution you'd like
Encode a fixed signature (expressed in the base64-encoded secret) in Azure function secrets. Encode a checksum in the final four bytes of the decoded data. Provide both mechanisms without compromising required security (bits of entropy) in the generated secret. 

GitHub's post on its [new authentication format](https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/) provides a good description of approach. Note that Azure DevOps, NuGet, and NPM also have implemented similar technique to provide additional security value.

#### Describe alternatives you've considered
Some providers have exclusively utilized a fixed signature (often a prefix) in the encoded value. This approach is subject to collisions, however, if there is a single (unrelated) data generator that duplicates the signature.

### PR DESCRIPTION
**Pull request by michaelcfanning**
### Issue describing the changes in this PR
Resolves #7662

### Pull request checklist

* [X] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [X] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [X] I have added all required tests (Unit tests, E2E tests)

@liliankasem @fabiocav @brettsam @mathewc 

Will send some additional context offline.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj [hunk 1]
```diff
@@ -70,6 +70,7 @@
     <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.32" />
     <PackageReference Include="Microsoft.Azure.WebJobs.Host.Storage" Version="5.0.0-beta.1" />
     <PackageReference Include="Microsoft.Azure.WebSites.DataProtection" Version="2.1.91-alpha" />
+    <PackageReference Include="Microsoft.Security.Utilities" Version="1.1.0" />
```
Comment (michaelcfanning) [inline src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: >Utilities [](http://example.com/codeflow?start=50&length=9)

This package contains key generation logic (and tests) that have been integrated in several other Azure resource providers. Note there is a pending change in this package possibly relevant to your RP (to produce URL friendly base64-encoded keys) that we could consider to update this PR.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 2]
```diff
@@ -36,6 +38,24 @@ public class SecretManager : IDisposable, ISecretManager
         private string _repositoryClassName;
         private DateTime _lastCacheResetTime;
 
+        internal const string AzureFunctionsSignature = "AzFu";
+
+        // Checksum seed for master keys. The string transformation here results in a
+        // a seed that is versioned in the least significant bits (as a ulong):
+        // BitConverter.ToUInt64(Encoding.ASCII.GetBytes("Master00").Reverse().ToArray());
+        // If there is a need to version, use "Master01" as the next string literal.
+        internal const ulong MasterKeySeed = 0x4d61737465723030;
```
Comment (michaelcfanning) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: >MasterKeySeed [](http://example.com/codeflow?start=29&length=13)

An important question for the team is how to 'bucket' keys by generating each set with a unique seed. We'd typically do this in situations where response/remediation may take a different path. In general, a good strategy is to group all user-facing (customer handled) secrets with one seed, and two organize other generated keys (which are intended to be internal-facing only) with another.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 3]
```diff
@@ -254,7 +274,7 @@ public async Task<KeyOperationResult> SetMasterKeyAsync(string value = null)
                 if (value == null)
                 {
                     // Generate a new secret (clear)
-                    masterKey = GenerateSecret();
+                    masterKey = GenerateIdentifiableSecret(MasterKeySeed);
```
Comment (michaelcfanning) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:267]: >GenerateIdentifiableSecret [](http://example.com/codeflow?start=32&length=26)

Assuming we retain a distinct seed for master keys, we should consider a distinct helper GenerateMasterKey or similar, which would ensure that the master key seed is always utilized. I did not make this change, though, wanting to get feedback from you on whether the grouping I've proposed here, master, system, and function keys, makes sense.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 4]
```diff
@@ -608,17 +628,14 @@ private Key CreateKey(string name, string secret)
...
 
-        internal static string GenerateSecret()
+        internal static string GenerateIdentifiableSecret(ulong seed)
         {
-            using (var rng = RandomNumberGenerator.Create())
-            {
-                byte[] data = new byte[40];
-                rng.GetBytes(data);
-                string secret = Convert.ToBase64String(data);
+            string secret = IdentifiableSecrets.GenerateIdentifiableKey(seed, 40, AzureFunctionsSignature);
```
Comment (michaelcfanning) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: >GenerateIdentifiableKey [](http://example.com/codeflow?start=48&length=23)

This is the essential change. This helper will produce a base64-encoded key, seven bytes of which are dedicated to a fixed signature (when encoded) and a checksum. The remaining entropy of the key is sufficient to meet security standards. Alternately, we could increase the size of the generated key in order to preserve full 320 bits of entropy (as exists today).

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 5]
```diff
@@ -608,17 +628,14 @@ private Key CreateKey(string name, string secret)
...
 
-                // Replace pluses as they are problematic as URL values
-                return secret.Replace('+', 'a');
-            }
+            secret = secret.Replace('+', '-');
```
Comment (michaelcfanning) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: >secret [](http://example.com/codeflow?start=12&length=6)

Previously, this code eliminated all '+' chars from the generated key, I believe because in at least one code path, for extensions, a key was returned as part of a URL. Generally, the '/' character is also replaced or eliminated if URL-friendly base64-encoded values are sought (and the '=' sign padding is also replaced.

What I've done here is simply to replace '+' with '-' and '/' with '_', a common convention for maintaining a base64-encoding with a complete character set.

This change, however, will change the alphabet of your keys moving forward. It will break any consumers who may have depended on the previous behavior to use functions like Convert.FromBase64String to convert a secret to a byte array.

The Azure Identity package provides an alternate which could now be used, the Base64UrlEncoder class. It would be important for this change to determine whether there are any code locations which s/be updated to use this API. Alternately, those code locations could simply find/replace back to the 'special' character set before calling Convert.

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs [hunk 6]
```diff
@@ -54,6 +55,88 @@ public SecretManagerTests(ITestOutputHelper outputHelper)
             _startupContextProvider = new StartupContextProvider(_testEnvironment, loggerFactory.CreateLogger<StartupContextProvider>());
         }
 
+        [Fact]
+        public void SecretManager_GenerateIdentifiableSecret_HonorsSeed()
```
Comment (michaelcfanning) [inline test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: >SecretManager_GenerateIdentifiableSecret_HonorsSeed [](http://example.com/codeflow?start=20&length=51)

I've been conservative about updating tests until we could have a preliminary conversation on the key grouping (i.e., the # of seeds used to bucket allocations). Once that's decided, several of these tests could be updated to exercise the key generation logic and to verify that populated values have the expected signature/checksum.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 2 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: In our case, all of these keys are user-facing - the keys just have different authorization levels. So the severity of an issue would vary based on key type. However, in all cases I think the response/remediation would be the same. So I'm not sure if we need the three buckets. @brettsam / @fabiocav what do you think?

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 5 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: As discussed, our keys need to continue to be URL friendly. They are passed as query params, but not in path segments which is why '/' isn't an issue. Also, we don't have any customers decoding these back to byte array - that's not a concern.

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj (hunk 1 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: What nuget feed is this on? Looks like on nuget.org the name is just reserved https://www.nuget.org/packages/Microsoft.Security.Utilities/

File: test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json [hunk 7]
```diff
@@ -6,7 +6,7 @@
   "compilationOptions": {},
   "targets": {
     ".NETCoreApp,Version=v6.0": {
-      "Microsoft.Azure.WebJobs.Script.WebHost/4.1.0": {
+      "Microsoft.Azure.WebJobs.Script.WebHost/4.1.1": {
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: Not sure if all these deps file changes are needed - we'll need to carefully review this.

Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: Yes, this jumped out to me as worrisome as well. Ideally, this file should only inject references specific to the sole dependency I've introduced, Microsoft.Security.Utilities. However, a specific unit test failed on my machine making that focused change, and so I went ahead and baselined. 

Several other tests fail as well, so my best bet is probably to consult with any machine/testing setup guidance you have to get the broader functionals working. One test appears to indicate Java be installed. Others require a missing connection string.

I'll be giving this PR attention tomorrow (to fix the API factoring improvements we discussed offline) and I'll reset this deps.json file. If you have a suggestion on ensuring all tests pass, let me know, otherwise will go digging around for info in the repo.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 5 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: Makes sense. Only the '+' character is a problem if tokens are only passed as query params. The trailing '=' padding chars shouldn't be a concern (as far as parameter parsing) if you're using the '&' as a separator.

Still, I think converting to a full 64-character (URL friendly) alphabet and eliding the trailing equal signs characters is a good change, since you've confirmed there shouldn't be anyone decoding to byte arrays (using standard base64) today.

One minor question I have is whether the absence of trailing padding might result in some incremental change in the # of false positive regex matches (that will get eliminated by the checksum validation). I plan to test this on a fairly large data set (a body of public OSS code) this week, will follow up here on results.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: The bucketing is most helpful for routing problems to the appropriate audience for response. If all keys are user-facing, you could definitely get by with one.

There is arguably a benefit in response, however, in having a more granular categorization, simple because more information in reporting maximizes actionability. i.e., if the distinction between master, system, and function keys in any ways maximizes motivation to fix, lowers costs to get oriented and respond, could be worth it. Certainly the scanning costs aren't high for bucketing.

My personal opinion is that it's a close call but I'd favor having more information/retaining the three buckets. But I wouldn't argue if you'd prefer to keep it simple.

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj (hunk 1 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: Can you see this version? https://www.nuget.org/packages/Microsoft.Security.Utilities/1.2.0. This is the one I plan to use.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: @mathewc , removed all deps.json changes (resulted from me baselining a debug build, I believe). will work on getting all functional tests running today.

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj (hunk 1 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: We just published v1.3 of this package, in order to provide a helper that generates a URL safe token without eliding the equal sign padding. We completed some offline analysis that suggests this will improve the accuracy of the preliminary regex match.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 8]
```diff
@@ -608,17 +642,10 @@ private Key CreateKey(string name, string secret)
...
-            {
-                byte[] data = new byte[40];
-                rng.GetBytes(data);
-                string secret = Convert.ToBase64String(data);
-
-                // Replace pluses as they are problematic as URL values
-                return secret.Replace('+', 'a');
-            }
+            string secret = IdentifiableSecrets.GenerateBase64Key(seed, 40, AzureFunctionsSignature, encodeForUrl: true);
+            return secret.EndsWith("==") ? secret : secret + "==";
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: Why the suffix logic "=="?

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs [hunk 9]
```diff
@@ -16,7 +16,9 @@
 using Microsoft.Azure.WebJobs.Script.Diagnostics;
 using Microsoft.Azure.WebJobs.Script.WebHost.Properties;
 using Microsoft.Extensions.Logging;
-using DataProtectionCostants = Microsoft.Azure.Web.DataProtection.Constants;
+using Microsoft.Security.Utilities;
+
+using DataProtectionConstants = Microsoft.Azure.Web.DataProtection.Constants;
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:21]: thanks for fixing the spelling :)

File: test/WebJobs.Script.Tests/DefaultScriptWebHookProviderTests.cs [hunk 10]
```diff
@@ -18,38 +19,68 @@ namespace Microsoft.Azure.WebJobs.Script.Tests
...
+                    "testextension_extension",
+                    It.IsAny<string>(),
+                    HostKeyScopes.SystemKeys,
+                    ScriptSecretsType.Host))
+                        .Callback<string, string, string, ScriptSecretsType>((key, secret, scope, type) => secretValue = secret)
+                        .Returns(() => Task.FromResult(new KeyOperationResult(secretValue, OperationResult.Created)));
+
+            // When an extension has no existing secret, one should be generated using
+            // the Azure Functions system key seed and standard fixed signature.
+            hostSecrets.SystemKeys = new Dictionary<string, string>();
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/DefaultScriptWebHookProviderTests.cs]: What does this line exist for? Doesn't look like these key values are accessed anywhere below?

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs [hunk 11]
```diff
@@ -1105,7 +1188,7 @@ private Mock<IKeyValueConverterFactory> GetConverterFactoryMock(bool simulateWri
             return mockValueConverterFactory;
         }
 
-        private SecretManager CreateSecretManager(string secretsPath, ILogger logger = null, IMetricsLogger metricsLogger = null, IKeyValueConverterFactory keyConverterFactory = null, bool createHostSecretsIfMissing = false, bool simulateWriteConversion = true, bool setStaleValue = true)
+        internal SecretManager CreateSecretManager(string secretsPath, ILogger logger = null, IMetricsLogger metricsLogger = null, IKeyValueConverterFactory keyConverterFactory = null, bool createHostSecretsIfMissing = false, bool simulateWriteConversion = true, bool setStaleValue = true)
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Security/SecretManagerTests.cs]: Is this change necessary? Didn't see it being accessed from other classes?

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs [hunk 12]
```diff
@@ -54,6 +55,88 @@ public SecretManagerTests(ITestOutputHelper outputHelper)
...
+                // Because we don't initialize any disk context, we can configure
+                // the secret manager to allocate fresh keys on start-up.
+                using (var secretManager = CreateSecretManager(directory.Path, createHostSecretsIfMissing: true))
+                {
+                    var hostSecrets = await secretManager.GetHostSecretsAsync();
+
+                    // Here and elsewhere we elide the '!' character which is prepended to every
+                    // generated secret value, as a clue that actual encryption is simulated in
+                    // the test case.
+                    ValidateSecret(hostSecrets.MasterKey.Substring(1), SecretManager.MasterKeySeed);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:84]: I don't understand this - who adds the "!"?

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs [hunk 13]
```diff
@@ -54,6 +55,88 @@ public SecretManagerTests(ITestOutputHelper outputHelper)
...
+                    }
+                }
+            }
+        }
+
+        private static void ValidateSecret(string secret, ulong seed)
+        {
+            // All Azure function keys are 40 bytes in length, 56 chars
+            // when base64-encoded. A URL friendly encoding elides the
+            // '==' padding, however.
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Security/SecretManagerTests.cs]: I don't understand this comment - SecretManager.GenerateIdentifiableSecret is appending == when not present.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 2 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: I'm fine with retaining the buckets, thanks.

Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: How about for these three seeds we add SecretManager unit tests that include the code you give and serve as the location where we can keep these comments and sample code. The tests run and are debuggable, modifiable for the future if we need to version. We could have just a simple comment referencing the tests for details?

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs (hunk 6 above)
Comment (mathewc) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: Agree that it would be good to update a few of the existing tests that add/update keys by adding a verification step at the end ensuring the right seed was used, the secrets are identifiable, etc. I.e. something that would catch a mistake if someone used the wrong seed value.

Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: Yes, makes sense. I will first merge this branch from main and then improve some tests. Are we all agreed that we're keeping the 3-seed breakdown, then? Master, system, and function?

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 8 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs]: Right, so we ran some tests and the absence of this trailing padding results in a significant number of matches against your core pattern (at GitHub public OSS scale). Adding the trailing padding will go a long way to reduce false positives for this preliminary match. We took a change to Security.Utilities to bring this configurability (padding or no padding) for URL friendly tokens into the API. I will update this PR package version and consume that new API.

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs (hunk 6 above)
Comment (mathewc) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: Yes, I'm fine with that. If @fabiocav / @brettsam don't have any opinions on this, I say go with what you have.

Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: ACK on this and your other notes, which I'm cleaning acting on today. Thanks for the advice.

File: test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json (hunk 7 above)
Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: I've reverted everything not relevant to MS.Security.Utilities.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs:46]: Great suggestion, done.

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs (hunk 11 above)
Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs]: Whoops, neglected to reset after deciding to add tests in-place.

File: test/WebJobs.Script.Tests/Security/SecretManagerTests.cs (hunk 6 above)
Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Security/SecretManagerTests.cs:61]: OK, so I have acted on your note, to update some tests to provide the validation you describe. I also made an effort to provide helpers that route people to the correct seeds.

**Review by mathewc — DISMISSED** [review]

File: test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json [hunk 14]
```diff
@@ -1322,6 +1323,14 @@
           }
         }
       },
+      "Microsoft.Security.Utilities/1.1.0": {
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: @fabiocav / @brettsam to help review these deps changes

Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: I just updated this deps file to properly reflect the current Microsoft.Security.Utilities dependency and the corresponding test passes. So just need to validate with everyone we are comfortable with picking up this Microsoft NuGet dependency.

File: test/WebJobs.Script.Tests/DefaultScriptWebHookProviderTests.cs (hunk 10 above)
Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/DefaultScriptWebHookProviderTests.cs]: This code was added to account for the fact that the working system requires a non-null dictionary for this property. I went ahead and just baked this into the default initialization of the mock object (in the ctor).

File: test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json (hunk 14 above)
Comment (michaelcfanning) [reply test/WebJobs.Script.Tests/Microsoft.Azure.WebJobs.Script.WebHost.deps.json]: @mathewc, FYI, I advertently cleared your previous approval, oops.

**Review by mathewc — DISMISSED** [review]

**Review by fabiocav — DISMISSED** [review]
Thank you for the changes and the patience, @michaelcfanning !

The changes look good. The only thing we need to do is make sure we're updating the runtime [assemblies manifest](https://github.com/Azure/azure-functions-host/blob/9cf4d03e54bcc98cc5b5e165467b56d04d0b6f99/src/WebJobs.Script/runtimeassemblies.json) to flag the new dependency ('Microsoft.Security.Utilities') as private, but I'm not going to block on this as I'd be happy to make that change if you'd like.

**Review by fabiocav — APPROVED** [review]

**Review by mathewc — APPROVED** [review]

**Comment by michaelcfanning** [comment]
Thanks, @fabiocav. I would, in fact, appreciate your help marking the new dependency as private. :) Thanks for that.

**Comment by mathewc** [comment]
Merged in mirror PR https://github.com/Azure/azure-functions-host/pull/8171
